src/probnum/linalg/solvers/stopping_criteria/_residual_norm.py

```diff
@@ -0,0 +1,40 @@
+"""Stopping criterion based on the norm of the residual."""
+
+import numpy as np
+
+import probnum
+from probnum.typing import ScalarArgType
+
+from ._linear_solver_stopping_criterion import LinearSolverStoppingCriterion
+
+
+class ResidualNormStopCrit(LinearSolverStoppingCriterion):
+    r"""Residual stopping criterion.
+
+    Terminate when the euclidean norm of the residual :math:`r_{i} = A x_{i} - b` is
+    sufficiently small, i.e. if it satisfies :math:`\lVert r_i \rVert_2 \leq \max(
+    \text{atol}, \text{rtol} \lVert b \rVert_2)`.
+
+    Parameters
+    ----------
+    atol :
+        Absolute tolerance.
+    rtol :
+        Relative tolerance.
+    """
+
+    def __init__(
+        self,
+        atol: ScalarArgType = 10 ** -5,
+        rtol: ScalarArgType = 10 ** -5,
+    ):
+        self.atol = probnum.utils.as_numpy_scalar(atol)
+        self.rtol = probnum.utils.as_numpy_scalar(rtol)
+
+    def __call__(
+        self, solver_state: "probnum.linalg.solvers.ProbabilisticLinearSolverState"
+    ) -> bool:
+
+        residual_norm = np.linalg.norm(solver_state.residual, ord=2)
+        b_norm = np.linalg.norm(solver_state.problem.b, ord=2)
```

Comment:
We might want to cache this

Reply:
As we previously discussed offline, I would avoid caching the residual or right-hand side at this point, before we actively need it for performance. We may want to either recompute the residual in each step for stability or update the (estimate of) right hand side at each step in the noisy case. Let me know if your opinion on this is different.